pkg/services/opsgenie/opsgenie_json.go

```diff
@@ -0,0 +1,129 @@
+//	})
+type AlertPayload struct {
+	Message     string            `json:"message"`
+	Alias       string            `json:"alias,omitempty"`
+	Description string            `json:"description,omitempty"`
+	Responders  []Entity          `json:"responders,omitempty"`
+	VisibleTo   []Entity          `json:"visibleTo,omitempty"`
+	Actions     []string          `json:"actions,omitempty"`
+	Tags        []string          `json:"tags,omitempty"`
+	Details     map[string]string `json:"details,omitempty"`
+	Entity      string            `json:"entity,omitempty"`
+	Source      string            `json:"source,omitempty"`
+	Priority    string            `json:"priority,omitempty"`
+	User        string            `json:"user,omitempty"`
+	Note        string            `json:"note,omitempty"`
+}
+
+// TODO: Duplicated code, see: formatter.go
+func deserializeSlice(str string) []string {
+	return strings.Split(str, ",")
+}
+
+// TODO: Duplicated code, see: formatter.go
+func deserializeMap(str string) (map[string]string, error) {
+	result := make(map[string]string)
+
+	pairs := strings.Split(str, ",")
+	for _, pair := range pairs {
+		elems := strings.Split(pair, ":")
+		if len(elems) != 2 {
+			return result, fmt.Errorf("field format is not supported")
+		}
+		key := elems[0]
+		value := elems[1]
+
+		result[key] = value
+	}
+
+	return result, nil
+}
+
+// NewAlertPayload instantiates AlertPayload
+func NewAlertPayload(message string, config *Config, params *types.Params) (AlertPayload, error) {
+	if params == nil {
+		params = &types.Params{}
+	}
+
+	result := AlertPayload{
+		Message: message,
+		// Populate with values from the query string as defaults
+		Alias:       config.Alias,
+		Description: config.Description,
+		Responders:  config.Responders,
+		VisibleTo:   config.VisibleTo,
+		Actions:     config.Actions,
+		Tags:        config.Tags,
+		Details:     config.Details,
+		Entity:      config.Entity,
+		Source:      config.Source,
+		Priority:    config.Priority,
+		User:        config.User,
+		Note:        config.Note,
+	}
+
+	// Populate with values from runtime parameters
+	for key, value := range *params {
+		var err error
+
+		switch key {
+		case "alias":
+			result.Alias = value
+		case "description":
+			result.Description = value
+		case "responders":
+			result.Responders, err = deserializeEntities(value)
+		case "visibleTo":
+			result.VisibleTo, err = deserializeEntities(value)
+		case "actions":
+			result.Actions = deserializeSlice(value)
+		case "tags":
+			result.Tags = deserializeSlice(value)
+		case "details":
+			result.Details, err = deserializeMap(value)
+		case "entity":
+			result.Entity = value
+		case "source":
+			result.Source = value
+		case "priority":
+			result.Priority = value
+		case "user":
+			result.User = value
+		case "note":
+			result.Note = value
+		default:
+			return result, fmt.Errorf("unknown config key: %q", key)
+		}
+
+		if err != nil {
+			return result, err
+		}
+
+	}
+
+	return result, nil
+}
```

Comment:
The propkeyresolver can update the config from params (which means that the config should be cloned before doing this), but it will stop and return an error for the structs, so you can just update them first and remove them from the map:

```suggestion
func NewAlertPayload(message string, config *Config, params *types.Params) (AlertPayload, error) {
	if params == nil {
		params = &types.Params{}
	}

	pkr := format.NewPropKeyResolver(config)

	if value, found := (*params)["responders"]; found {
		responders, _ := deserializeEntities(value)
		config.Responders = responders
		delete(*params, "responders")
	}

	if value, found := (*params)["visibleTo"]; found {
		responders, _ := deserializeEntities(value)
		config.Responders = responders
		delete(*params, "visibleTo")
	}

	if err := pkr.UpdateConfigFromParams(config, params); err != nil {
		return AlertPayload{}, err
	}

	result := AlertPayload{
		Message: message,
		// Populate with values from the query string as defaults
		Alias:       config.Alias,
		Description: config.Description,
		Responders:  config.Responders,
		VisibleTo:   config.VisibleTo,
		Actions:     config.Actions,
		Tags:        config.Tags,
		Details:     config.Details,
		Entity:      config.Entity,
		Source:      config.Source,
		Priority:    config.Priority,
		User:        config.User,
		Note:        config.Note,
	}

	return result, nil
}
```

Reply:
Brilliant, that makes everything simpler. I've added it with some minor moidifications. Not sure what you mean by

> Doing the param update should probably be moved to opsgenie.go since Service already has a PropKeyResolver instance ready to use.

I couldn't find a `PropKeyResolver` in `opsgenie.go`. There is a transient one in `opsgenie_config.go` that I could potentially save and re-use.

Reply:
No, sorry, usually there is an instance of `PropKeyResolver` stored on the `Service` struct for the services. That's because it's used at least once (often twice) in initializing, for setting the config defaults and for updating the config according to the supplied URL. 
Example:
https://github.com/containrrr/shoutrrr/blob/main/pkg/services/telegram/telegram.go#L30-L56

Reply:
Got it, fixed